### PR TITLE
Support TypeScript, multiple configs, other spring cleaning

### PR DESCRIPTION
**Pull request by mackenco**
This is a big overhaul of our ESLint package. I'm leaning a bit on [Kent Dodds's ESLint packages](https://github.com/kentcdodds/eslint-config-kentcdodds) which I think are structured in a relatively sane way for sharing across projects. I tested this out in Warbler, Shrike, and Flex Funnels. 

## Big Changes
* I've broken the config out in to two configs - a base config, and a `react` flavor, which is the base + React and jsx-a11y rules. Using the config will now look like this:
```js
// !React repos
module.exports = {
  extends: 'arcadia',
};

// React apps
module.exports = {
  extends: 'arcadia/react',
};
```
* The configs account for TypeScript, and extend some `@typescript-eslint` rules.

## (Bill Maher voice) New Rules
* Maybe the most controversial thing in this PR, I turned off [`sort-keys`](https://eslint.org/docs/rules/sort-keys) because it drives me nuts. 
* I turned on [`@typescript-eslint/explicit-function-return-type`](https://typescript-eslint.io/rules/explicit-function-return-type/) and [`@typescript-eslint/ban-ts-comment`](https://typescript-eslint.io/rules/ban-ts-comment/) (unless there is a `description`). In my testing, these (especially `explicit-function-return-type`) throw a lot of errors. I think these are rules we want in general, but I think we'll have to be okay with adopters turning them off for now? 
* I turned off [`react/function-component-definition`](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/function-component-definition.md) and [`react/jsx-curly-spacing`](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-curly-spacing.md). When I was testing this out in FF, neither of those were being followed at all. 
* I added this nifty `eslint-find-rules` package, which surfaces rules that exist that we're not accounting for. It surfaced a few, I added them with the Kent Dodds defaults: 
  * [`prefer-object-has-own`](https://eslint.org/docs/rules/prefer-object-has-own) - error
  * [`react/hook-use-state`](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/hook-use-state.md) - off
  * [`react-hooks/rules-of-hooks`](https://github.com/facebook/react/tree/main/packages/eslint-plugin-react-hooks) - error
  * [`react-hooks/exhaustive-deps`](https://github.com/facebook/react/tree/main/packages/eslint-plugin-react-hooks) - warn

## Small Changes
* Added an `.nvmrc` and `.gitignore`
* Added Husky + a `pre-commit` hook which lints and checks if there are any missing rules. 

## TODO / Questions / Bikeshedding topics
* Publishing - publishing on this is still manual; would be nice to automate (could be a good candidate for Github packages?)
* Renaming - should we rename this to `@arcadia-power/eslint-config`?? Would it be crazy to start using the namespace, or should we do that all in one big push
* Rules - any more rules people want tweaked? As long as we're cutting a new major, now would be a good time
* AirBnB - would we ever stop using AirBnB? It'd be kind of gnarly, but I'm not opposed to copying over all the rules + AirBnB defaults into this repo, just so we "own" everything and it's all in one place. 
* Testing config - do we want to do a Jest/testing config as well? https://github.com/kentcdodds/eslint-config-kentcdodds/blob/main/jest.js
* TypeScript - do we want to turn on the ["recommended" TS rules](https://github.com/typescript-eslint/typescript-eslint/tree/main/packages/eslint-plugin#recommended-configs)? It will probably make adopting this harder for TS projects (as they'll start to get a lot of errors), but seems like things could be fixed/ignored as needed, and this will get us in a better place going forward as we write TS?

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,121 +1,73 @@
...
+    'no-shadow': 'off',
+    'no-underscore-dangle': 'off',
+    'no-use-before-define': 'off',
+    'object-curly-newline': 'off',
+    'prefer-arrow-callback': 'off',
+    'prefer-destructuring': 'error',
+    'prefer-object-has-own': 'error',
+    'prefer-rest-params': 'off',
+    radix: 'off',
+    'sort-keys': 'off',
```
Comment (harkshark) [inline index.js:44]: I feel shocked

File: package.json [hunk 2]
```diff
@@ -1,19 +1,46 @@
...
+  "description": "Arcadia's ESLint Configurations",
   "main": "index.js",
-  "author": "Arcadia Power Engineering",
+  "author": "Arcadia Engineering",
   "license": "ISC",
   "keywords": [
     "eslint"
   ],
+  "scripts": {
+    "find-new-rules": "npm-run-all --parallel find-new-rules:*",
```
Comment (harkshark) [inline package.json:12]: Neat

File: package.json [hunk 3]
```diff
@@ -1,19 +1,46 @@
 {
   "name": "eslint-config-arcadia",
```
Comment (harkshark) [inline package.json:2]: >Should we rename this to @arcadia-power/eslint-config?? Would it be crazy to start using the namespace, or should we do that all in one big push

What _else_ would be included in a "big push"? Are you thinking platform repos?

File: Readme.md [hunk 4]
```diff
@@ -1,26 +1,66 @@
...
+  [`eslint-plugin-jsx-a11y`](https://github.com/evcohen/eslint-plugin-jsx-a11y)
+  for rules regarding accessibility with JSX (needed in the `react` config)
+- `@typescript-eslint`:
+  [@typescript-eslint/parser](https://www.npmjs.com/package/@typescript-eslint/parser)
+  and
+  [@typescript-eslint/eslint-plugin](https://www.npmjs.com/package/@typescript-eslint/eslint-plugin) (needed for projects written in TypeScript)
 
 ## Publishing
+
 The config is hosted as an NPM package on Gemfury.
```
Comment (harkshark) [inline Readme.md:64]: >Publishing on this is still manual; would be nice to automate (could be a good candidate for Github packages?)

+1 on this a potential first candidate for GitHub Actions! Seems more low-risk than Shrike or Warbler!

File: package.json (hunk 3 above)
Comment (mackenco) [reply package.json:2]: No, I'm thinking namespacing our other JS NPM packages (shrike, warbler)

File: package.json [hunk 5]
```diff
@@ -1,19 +1,46 @@
...
+    "react": "^18.0.0",
+    "typescript": "^4.6.3"
+  },
   "peerDependencies": {
-    "babel-eslint": ">=8.2.3",
-    "eslint": ">=4",
-    "eslint-config-airbnb": ">=16.1.0",
-    "eslint-plugin-react": ">=7.8.2",
-    "eslint-plugin-import": ">=2.7.0",
-    "eslint-plugin-jsx-a11y": ">=6.0.2"
```
Comment (twdrake-ap) [inline package.json:17]: 🙌

File: index.js (hunk 1 above)
Comment (twdrake-ap) [reply index.js]: 👍

**Review by twdrake-ap — APPROVED** [review]
Looks good to me, I say get this released and we can iterate on it as it gets adopted.

**Pull request merged**